scalalib/src/mill/javalib/android/AndroidAppBundle.scala

```diff
@@ -0,0 +1,214 @@
+
+    // Get resource directories and filter for existing ones
+    val resourceDirs = resources().map(_.path).filter(os.exists)
+
+    // Download and extract AAPT2 tool
+    os.write(aapt2Jar, requests.get(aapt2Url()).bytes)
+    os.unzip(aapt2Jar, Task.dest)
+    os.call(("chmod", "+x", aapt2)) // Make AAPT2 executable
+    os.makeDir.all(compiledResDir) // Ensure compiled resource directory exists
+
+    // Compile each resource directory using AAPT2
+    var count = 0
+    val compiledZips = resourceDirs.map { resDir =>
+      val outputZip = compiledResDir / s"${resDir.last}-${count}.zip"
+      count += 1
+      os.call((
+        aapt2,
+        "compile",
+        "--dir",
+        resDir,
+        "-o",
+        outputZip
+      ))
+      outputZip
+    }
+
+    // Separate main resource zip and library zips for later linking
+    val ResourceZip = compiledZips.find(_.toString.contains("resources"))
+    val libzips = compiledZips.filterNot(_.toString.contains("resources"))
+    val compiledlibs = libzips.flatMap(zip => Seq("-R", zip.toString))
+
+    // Link resources and generate res.apk
+    os.call(
+      Seq(
+        aapt2.toString, // AAPT2 executable path
+        "link",
+        "-I",
+        androidSdkModule().androidJarPath().path.toString, // Android SDK jar
+        "--auto-add-overlay",
+        "--proto-format" // Format and overlay options
+      ) ++ compiledlibs ++ Seq(
+        "--manifest",
+        androidManifest().path.toString, // Use provided manifest
+        "--java",
+        Task.dest.toString, // Output generated R.java
+        "-o",
+        s"${Task.dest / "res.apk"}", // Output APK path
+        ResourceZip.map(_.toString).getOrElse("")
+      )
+    )
+
+    PathRef(Task.dest) // Return compiled resource directory as PathRef
+  }
+
+  /**
+   * Packages compiled resources and dex files into a zip for app bundle generation.
+   *
+   * @return PathRef to the created bundle zip.
+   */
+  def androidBundleZip: T[PathRef] = Task {
+    val dexFile = androidDex().path
+    val baseDir = Task.dest / "base"
+
+    // Create directory structure for Android bundle
+    Seq("manifest", "dex").foreach(d => os.makeDir.all(baseDir / d))
+
+    // Copy dex and resource files into the bundle directory
+    os.copy(dexFile, Task.dest / dexFile.last)
+    os.unzip(Task.dest / os.up / "androidResources.dest" / "res.apk", Task.dest)
+    os.copy(Task.dest / "AndroidManifest.xml", baseDir / "manifest" / "AndroidManifest.xml")
+
+    // Copy .pb files (compiled protobuf resources) and organize dex files
+    os.walk(Task.dest).filter(_.ext == "pb").foreach(file => os.copy(file, baseDir / file.last))
+    os.walk(Task.dest).filter(_.ext == "dex").zipWithIndex.foreach { case (file, idx) =>
+      os.copy(file, baseDir / "dex" / (if (idx == 0) "classes.dex" else s"classes${idx + 1}.dex"))
+    }
+
+    // Copy additional resources (res, lib, assets) if present
+    Seq("res", "lib", "assets").foreach { folder =>
```

Comment:
Better to use `for` loops than `.foreach` for this kind of thing, here and elsewhere in the PR